cmd/goat/account_migrate.go

```diff
@@ -0,0 +1,263 @@
+		Did:        &did,
+		Handle:     newHandle,
+		Password:   &newPassword,
+	}
+	if newEmail != "" {
+		createParams.Email = &newEmail
+	}
+	if inviteCode != "" {
+		createParams.InviteCode = &inviteCode
+	}
+
+	// use service auth for access token, temporarily
+	newClient.Auth = &xrpc.AuthInfo{
+		Did: did,
+		Handle: newHandle,
+		AccessJwt: createAuthResp.Token,
+		RefreshJwt: createAuthResp.Token,
+	}
+	createAccountResp, err := comatproto.ServerCreateAccount(ctx, &newClient, &createParams)
+	if err != nil {
+		return fmt.Errorf("failed creating new account: %w", err)
+	}
+
+	if createAccountResp.Did != did {
+		return fmt.Errorf("new account DID not a match: %s != %s", createAccountResp.Did, did)
+	}
+	newClient.Auth.AccessJwt = createAccountResp.AccessJwt
+	newClient.Auth.RefreshJwt = createAccountResp.RefreshJwt
+
+	// login client on the new host
+	sess, err := comatproto.ServerCreateSession(ctx, &newClient, &comatproto.ServerCreateSession_Input{
+		Identifier: did,
+		Password:   newPassword,
+	})
+	if err != nil {
+		return fmt.Errorf("failed login to newly created account on new host: %w", err)
+	}
+	newClient.Auth = &xrpc.AuthInfo{
+		Did:        did,
+		AccessJwt:  sess.AccessJwt,
+		RefreshJwt: sess.RefreshJwt,
+	}
+
+	// 2. Migrate Data
+
+	slog.Info("migrating repo")
+	repoBytes, err := comatproto.SyncGetRepo(ctx, oldClient, did, "")
+	if err != nil {
+		return fmt.Errorf("failed exporting repo: %w", err)
+	}
+	err = comatproto.RepoImportRepo(ctx, &newClient, bytes.NewReader(repoBytes))
+	if err != nil {
+		return fmt.Errorf("failed importing repo: %w", err)
+	}
+
+	slog.Info("migrating preferences")
+	// TODO: service proxy header for app.bsky?
+	prefResp, err := appbsky.ActorGetPreferences(ctx, oldClient)
+	if err != nil {
+		return fmt.Errorf("failed fetching old preferences: %w", err)
+	}
+	err = appbsky.ActorPutPreferences(ctx, &newClient, &appbsky.ActorPutPreferences_Input{
+		Preferences: prefResp.Preferences,
+	})
+	if err != nil {
+		return fmt.Errorf("failed importing preferences: %w", err)
+	}
+
+	slog.Info("migrating blobs")
+	blobCursor := ""
+	for {
+		listResp, err := comatproto.SyncListBlobs(ctx, oldClient, blobCursor, did, 100, "")
+		if err != nil {
+			return fmt.Errorf("failed listing blobs: %w", err)
+		}
+		for _, blobCID := range listResp.Cids {
+			blobBytes, err := comatproto.SyncGetBlob(ctx, oldClient, blobCID, did)
+			if err != nil {
+				slog.Warn("failed downloading blob", "cid", blobCID, "err", err)
+				continue
+			}
+			_, err = comatproto.RepoUploadBlob(ctx, &newClient, bytes.NewReader(blobBytes))
+			if err != nil {
+				slog.Warn("failed uploading blob", "cid", blobCID, "err", err, "size", len(blobBytes))
+			}
+			slog.Info("transfered blob", "cid", blobCID, "size", len(blobBytes))
+		}
+		if listResp.Cursor == nil || *listResp.Cursor == "" {
+			break
+		}
+		blobCursor = *listResp.Cursor
+	}
+
+	// display migration status
+	// TODO: should this loop? with a delay?
```

Comment:
yeah you could run this in a loop in the background to give like progress bars for blob uploads or something, but otherwise you can just do it in between steps to ensure that things finished correctly